daemon/logger/amqp/amqp.go

```diff
@@ -0,0 +1,419 @@
+// +build linux
+
+package amqp
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/docker/daemon/logger"
+	"github.com/docker/docker/daemon/logger/loggerutils"
+	"github.com/streadway/amqp"
+)
+
+const name = "amqp"
+
+type amqpLogger struct {
+	ctx        logger.Context
+	fields     amqpFields
+	connection *amqpConnection
+}
+
+// Data structure holding information about the current connection
+// with a broker as well as a list of other available brokers
+type amqpConnection struct {
+	broker     int
+	brokerURLs []*amqpBroker
+	conn       *amqp.Connection
+	c          *amqp.Channel
+	conf       <-chan amqp.Confirmation
+	err        error
+}
+
+// Data structure to hold the connection settings for each broker
+type amqpBroker struct {
+	BrokerURL  *url.URL
+	Exchange   string
+	Queue      string
+	RoutingKey string
+	Tag        string
+	CertPath   string
+	KeyPath    string
+	Confirm    bool
+}
+
+// Data structure to store the data for the log message
+type amqpMessage struct {
+	Message   string     `json:"message"`
+	Version   string     `json:"@version"`
+	Timestamp time.Time  `json:"@timestamp"`
+	Tags      amqpFields `json:"tags"`
+	Host      string     `json:"host"`
+	Path      string     `json:"path"`
+}
+
+// Data about the host and container that is required when sending
+// the log message
+type amqpFields struct {
+	Hostname      string
+	ContainerID   string
+	ContainerName string
+	ImageID       string
+	ImageName     string
+	Command       string
+	Tag           string
+	AMQPTag       string
+	Created       time.Time
+}
+
+func init() {
+	if err := logger.RegisterLogDriver(name, New); err != nil {
+		logrus.Fatal(err)
+	}
+	if err := logger.RegisterLogOptValidator(name, ValidateLogOpt); err != nil {
+		logrus.Fatal(err)
+	}
+}
+
+// New creates a new amqp logger using the configuration passed in the
+// context.
+func New(ctx logger.Context) (logger.Logger, error) {
+	// collect extra data for AMQP message
+	hostname, err := ctx.Hostname()
+	if err != nil {
+		return nil, fmt.Errorf("Cannot access hostname to set source field: %v", err)
+	}
+
+	logrus.Infof("URLs: %v", ctx.Config["amqp-url"])
+
+	// remove trailing slash from container name
+	containerName := bytes.TrimLeft([]byte(ctx.ContainerName), "/")
+
+	tag, err := loggerutils.ParseLogTag(ctx, "")
+	if err != nil {
+		return nil, err
+	}
+
+	fields := amqpFields{
+		Hostname:      hostname,
+		ContainerID:   ctx.ContainerID,
+		ContainerName: string(containerName),
+		ImageID:       ctx.ContainerImageID,
+		ImageName:     ctx.ContainerImageName,
+		Command:       ctx.Command(),
+		Tag:           tag,
+		AMQPTag:       ctx.Config["amqp-tag"],
+		Created:       ctx.ContainerCreated,
+	}
+
+	broker := 0
```

Comment:
That makes sense. I was just being explicit here that the broker is initially set to 0.